### ISSUE TITLE
Apply Markdown and HTML styling

### ISSUE THREAD

**Issue by lognaturel** [getodk/web-forms#62]
Collect's implementation: https://github.com/getodk/collect/blob/e49eae16657d6677abde44e4475c1f08d3ccf9ae/collect_app/src/main/java/org/odk/collect/android/utilities/HtmlUtils.java#L23
- It includes some subtleties such as https://github.com/enketo/enketo/issues/24

Documentation: https://docs.getodk.org/form-styling/#markdown

Enketo allows disabling style: https://github.com/enketo/enketo/blob/37a7af1886743c7e20ddcb9f3b0f8708d184cd88/packages/enketo-transformer/README.md?plain=1#L85
* @eyelidlessness: "I think it’s something we might want, eg maybe for like summary views. Either way I think if we’re gonna parse it, that would happen in the engine, and then we tailor the representation to what clients need and what makes the most sense for all of the common concerns around markdown. Structured data feels much safer to me than serializing it in the engine, both in terms of security and client flexibility"

Possible subtleties around using dynamic values in links: https://github.com/enketo/enketo/issues/804

Consider using https://github.com/remarkjs/remark

**Comment by lognaturel** [comment]
@eyelidlessness here's a form: [style-example.xlsx](https://github.com/user-attachments/files/17035458/style-example.xlsx)

**Comment by garethbowen** [comment]
Jotting down my thoughts before I head off for the weekend and forget it all...

1. Don't forget to markdown translated text
2. TextChunks will have to be combined before formatting, because you might have `hello **${yourname}**!` which becomes three chunks but the open tag and close tags are in different chunks.
3. Escape output chunks so user entered strings don't mess with the formatting (either maliciously or accidentally).
4. [Option 1](https://github.com/getodk/web-forms/issues/198) is probably the right approach. This means the engine parses the markdown, and the client looks after the rendering. This puts more work on the client (parsing, sanitising, and all the bugs therein), but leaves the engine unopinionated, so it works with all sorts of clients (eg: scenario testing). Will need to make sure it's not too impactful on bundle size (à la tree-sitter).

**Comment by latin-panda** [comment]
@garethbowen, it would be helpful to test it in Central. We previously had issues where Central's bootstrap and other styles affected Web Forms' UI display

**Comment by lognaturel** [comment]
> Do we want to handle images in markdown?

Not currently. It's something that has come up before but we should treat it as a spec change, not something we add to this implementation. If we were to consider it, I think we'd want to carefully think through what user expectations when using a form offline should be (e.g. is a user warned when uploading a form definition that md images are only available online? Are clients expected to download and cache those images?).

> doesn't let the form builder define any old HTML they want

Ideally only `span`, `p`, `div` and HTML entities are rendered at all. I think in practice Collect does render more but what it does render is based on the Android version it's used on. If we wanted to open up to supporting more HTML I think we should consider that a spec change.

> I propose we allow this, as long as the only actor capable of injecting is the form builder and not the end user.

Collect renders markup in `outputs` and I believe the goal is to match that (e.g. https://github.com/getodk/web-forms/issues/117#issuecomment-3309493769). That would enable users to introduce arbitrary markup: [style.xlsx](https://github.com/user-attachments/files/22667284/style.xlsx).

Maybe in a first PR, HTML can continue to be passed through literally to the client and we can file a separate issue for handling HTML? Basically at this point we could tentatively say we consider HTML rendering and sanitizing a client responsibility and then we could design an approach for that. One idea is that our Vue client could provide an interface for the host application to provide a sanitization function of its choice. Central already uses `DOMPurify` and it could whitelist `span`, `p`, `div`, `style` for its WF needs. Another host application could choose to allow more HTML.

**Comment by lognaturel** [comment]
> Escape output chunks so user entered strings don't mess with the formatting

Missed this before! So far have you been thinking of this as escape forever -- styling in outputs is never applied (and then my comment above about possible user-provided HTML isn't right) -- or first apply styling for the containing string and then for outputs to avoid e.g. a `*` in an output value closing formatting that it wasn't supposed to be connected with?

I've seen some forms like the `jr:choice-name` one I shared previously with legitimate uses of markup in outputs but not many. We could certainly leave that behind for now and come back to it if it's explicitly requested. As you note, Enketo doesn't support this.

**Comment by garethbowen** [comment]
> So far have you been thinking of this as escape forever

Yes that's been my approach so far, as you say, to avoid user entered markdown closing the form defined markdown unintentionally.

> or first apply styling for the containing string and then for outputs

I quite like this and I think it's not too challenging. I'll look into it and see how I get on.

**Comment by garethbowen** [comment]
There's a styling issue if the question label has a block element (heading, list, div) **and** is required. The label gets out of alignment with the `*` firstly because block level elements take a line to themselves, but additionally might have additional padding/margin. For example:

<img width="836" height="264" alt="Image" src="https://github.com/user-attachments/assets/ee123b1e-5ac8-4cfc-bb79-f04b9744b32b" />

By floating the asterisk to the left the block level element is no longer pushed down onto a new line. This doesn't solve the padding/margin but improves things a little. 

<img width="836" height="264" alt="Image" src="https://github.com/user-attachments/assets/009c3fd2-cfdf-4f14-b684-b5a17ea46d63" />

We could go further by eliminating the top margin from the label's first child element which is quite heavy handed but does end up with the desired result.

<img width="836" height="264" alt="Image" src="https://github.com/user-attachments/assets/3ed73561-1526-44e4-b4aa-0f9a83aadd46" />

@alyblenkin What do you think? Is this too pedantic, or not pedantic enough?

**Comment by garethbowen** [comment]
> Will need to make sure it's not too impactful on bundle size...

Just tested this to make sure. I checked the web-forms/dist/index.js:

`main`: 3,363.30 kB │ gzip: 1,184.49 kB
`62-markdown`: 3,499.83 kB │ gzip: 1,218.56 kB

I think this is probably acceptable for what is a fairly chunky feature.
